snuba/query/data_source/join.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Generic, Mapping, NamedTuple, Optional, Sequence, Union
+
+from snuba.clickhouse.columns import ColumnSet, QualifiedColumnSet, SchemaModifiers
+from snuba.query import ProcessableQuery, TSimpleDataSource
+from snuba.query.data_source import DataSource
+
+
+class JoinType(Enum):
+    INNER = "INNER"
+    LEFT = "LEFT"
+
+
+class JoinModifier(Enum):
+    ANY = "ANY"
+    SEMI = "SEMI"
+
+
+@dataclass(frozen=True)
+class JoinNode(ABC, Generic[TSimpleDataSource]):
+    """
+    Assigns an alias to a node in a join expression.
+    The alias is used in the join condition and in all the expressions
+    that rely on the join in the query.
+
+    Join nodes can be simple data sources, join expressions and queries.
+    """
+
+    alias: Optional[str]
```

Comment:
Maybe this should be mandatory. I know that it's not technically needed all the time but it is sometimes required.

Reply:
It cannot, A join condition does not have an alias and it is a JoinNode, though thinking over this, I think I should be able to put this field only on Individual node.